### PR TITLE
Add tweet status column for supported post types

### LINKED ISSUE #56: Finding Tweeted posts
**Is your enhancement related to a problem? Please describe.**
Faster identification of posts which need Tweeting.  Supporting multi-editor workflow:
* Placement and persistence of Tweet message to support multi editor workflow
* Configurable representation of Tweet status in Post listing
* Configurable Tweet status filter

**Describe the solution you'd like**
Update the All Posts screen to:
- [ ] add filters for "Tweeted" and/or "Not Tweeted"
- [ ] add column for Tweeted status
- [ ] add hover state for rows in Tweeted status column to show the Tweet date(s) and time(s) in a simple title attribute

**Designs**
Screen Shot 2019-12-13 at 7 20 16 PM

**Describe alternatives you've considered**
n/a

**Additional context**
 Props to @linawiezkowiak and @oszkarnagy for the UX/Design work.

### PR DESCRIPTION
**Pull request by thrijith**
### Description of the Change

Adds changes outlined in https://github.com/10up/autoshare-for-twitter/issues/56

### Verification Process

- Enabled and activate the plugin, made changes for adding custom column in `feature/add-custom-filter`
- Tested post list for tweeted posts

Normal view
Screenshot 2021-02-02 at 11 11 13 PM

Hover view

Screenshot 2021-02-02 at 11 12 01 PM

### Checklist:

- [x] I have read the [**CONTRIBUTING**](/CONTRIBUTING.md) document.
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my change.
- [ ] All new and existing tests passed.

### Applicable Issues

Fixes https://github.com/10up/autoshare-for-twitter/issues/56

### Changelog Entry

Add Tweet status column for supported / enabled post types.

### CODE REVIEW

File: includes/core.php [hunk 1]
```diff
@@ -80,3 +84,73 @@ function maybe_disable_upload_image( $attachment_id ) {
...
+ */
+function modify_post_type_add_tweet_status_column( $columns ) {
+	// Do this so our custom column doesn't end up being the last one, messing up UI.
+	unset( $columns['date'] );
+
+	// Add tweet status column header.
+	$columns['is_tweeted'] = sprintf(
+		'<span class="autoshare-for-twitter-status-logo" title="%s"><span class="screen-reader-text">%s</span></span>',
+		esc_attr__( 'Tweeted status', 'autoshare-for-twitter' ),
+		esc_html__( 'Tweeted status', 'autoshare-for-twitter' )
```
Comment (dinhtungdu) [inline includes/core.php:102]: You can use `%1$s` here instead.
```suggestion
		'<span class="autoshare-for-twitter-status-logo" title="%2$s"><span class="screen-reader-text">%1$s</span></span>',
		esc_attr__( 'Tweeted status', 'autoshare-for-twitter' )
```

File: includes/core.php [hunk 2]
```diff
@@ -80,3 +84,73 @@ function maybe_disable_upload_image( $attachment_id ) {
...
+ */
+function modify_post_type_add_tweet_status( $column_name, $post_id ) {
+	if ( 'is_tweeted' === $column_name ) {
+		$post_status = get_post_status( $post_id );
+
+		$tweet_status = Utils\get_autoshare_for_twitter_meta( $post_id, TWITTER_STATUS_KEY );
+		$status       = isset( $tweet_status['status'] ) ? $tweet_status['status'] : '';
+
+		if ( 'publish' !== $post_status || empty( $status ) ) {
+			echo '';
```
Comment (dinhtungdu) [inline includes/core.php]: This `echo` call can be eliminated by check for the opposite conditions.

File: includes/core.php [hunk 3]
```diff
@@ -80,3 +84,73 @@ function maybe_disable_upload_image( $attachment_id ) {
...
 }
+
+/**
+ * Adds tweeted status columns for supported post types.
+ */
+function add_tweeted_column() {
+	// Loop through all the supported post types and add tweet status column.
+	$post_types = Utils\get_enabled_post_types();
+	foreach ( (array) $post_types as $post_type ) {
+		add_post_type_support( $post_type, POST_TYPE_SUPPORT_FEATURE );
```
Comment (dinhtungdu) [inline includes/core.php]: this duplicates the functionalities of `set_post_type_supports()`

File: includes/core.php [hunk 4]
```diff
@@ -80,3 +84,73 @@ function maybe_disable_upload_image( $attachment_id ) {
...
+				$twitter_url = Utils\link_from_twitter( $tweet_status['twitter_id'] );
+				$tweet_title = sprintf(
+					'%s %s',
+					__( 'Tweeted on', 'autoshare-for-twitter' ),
+					$date
+				);
+
+				printf(
+					'<a href="' . esc_url( $twitter_url ) . '" target="_blank">
+					<span class="autoshare-for-twitter-status-logo" title="' . esc_attr( $tweet_title ) . '"></span>
```
Comment (dinhtungdu) [inline includes/core.php]: `title` attribute should be added to `<a>` instead.

File: includes/core.php [hunk 5]
```diff
@@ -80,3 +84,73 @@ function maybe_disable_upload_image( $attachment_id ) {
...
+}
+
+/**
+ * Add tweet status data to each row.
+ *
+ * @param  string $column_name Column name.
+ * @param  int    $post_id Post ID.
+ */
+function modify_post_type_add_tweet_status( $column_name, $post_id ) {
+	if ( 'is_tweeted' === $column_name ) {
```
Comment (dinhtungdu) [inline includes/core.php]: Using early return style can save us up to 2 indent levels.

**Review by dinhtungdu — CHANGES_REQUESTED** [review]
Hi @thrijith, thank you for your PR, it looks good to me. However, there are some minor issues in your PR, can you please fix them to get this ready to be merged:

- The post filter hasn't been added. I notice you set this PR as draft, so you may be working on it.
- The tweeted icon in the table header shouldn't have a hover state.
- The tweeted icon doesn't really match the one in the design to me, [this one](https://www.freepik.com/free-icon/twitter-logo_748486.htm) looks closer.
- When you click on the tweeted icon (which opens the tweet page in a new tab), there are two blue dots on the left of the clicked icon.

File: includes/core.php (hunk 5 above)
Comment (thrijith) [reply includes/core.php]: @dinhtungdu updated, please check again.

File: includes/core.php (hunk 4 above)
Comment (thrijith) [reply includes/core.php]: @dinhtungdu updated, please check again.

File: includes/core.php (hunk 3 above)
Comment (thrijith) [reply includes/core.php]: @dinhtungdu I've moved similar code to the other int hook function we already had.

File: includes/core.php (hunk 2 above)
Comment (thrijith) [reply includes/core.php]: @dinhtungdu updated, please check again.

**Comment by thrijith** [comment]
> The post filter hasn't been added. I notice you set this PR as draft, so you may be working on it.

Yes, currently checking for options to do the same ref https://github.com/10up/autoshare-for-twitter/issues/56#issuecomment-773611059

> The tweeted icon in the table header shouldn't have a hover state.

Done.

> The tweeted icon doesn't really match the one in the design to me, [this one](https://www.freepik.com/free-icon/twitter-logo_748486.htm) looks closer.

Unable to download the linked one, we need two for each hover and inactive state, not getting the issue with current icons.

> When you click on the tweeted icon (which opens the tweet page in a new tab), there are two blue dots on the left of the clicked icon.

Done.

File: includes/core.php (hunk 1 above)
Comment (thrijith) [reply includes/core.php:102]: @dinhtungdu One value is used in the attributes while other is inside the tag, so added translation functions, please let me know if that needs change, thanks!

**Review by dinhtungdu — APPROVED** [review]

**Pull request merged**